plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1256,6 +1262,22 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
                 []    -> a
                 _ : _ -> b
           {-# INLINE choosePlc #-}
+    toBuiltinMeaning _ver CaseList =
+        makeBuiltinMeaning
+            caseListPlc
+            (\_ _ _ _ -> ExBudgetLast mempty)  -- TODO.
```

Comment:
I was wondering what happens to the cost...

Reply:
We always do a constant amount of work, so should be constant and not very hard to measure. One part that may be tricky is distinguishing (costing-wise) between returning a function application from a builtin and evaluating that application in the CEK machine. But we probably can adapt the current Kenneth's trick of using dedicated no-op builtins. Anyways, we should only discuss it if we think pattern matching builtins make sense, otherwise we'll just waste a bunch of time. I do believe costing shouldn't be an issue, but I might be wrong about that.

Reply:
That's good to know. I assumed so, but when I asked why there is no higher-order builtin (especially for cases like the ones here where higher-order is the most natural formulation) I was told that it was because of the costs, so I was wondering if there is any problem here.

Reply:
> That's good to know. I assumed so, but when I asked why there is no higher-order builtin (especially for cases like the ones here where higher-order is the most natural formulation) I was told that it was because of the costs

We _can_ define higher-order builtins already and we _can_ cost the definable ones properly. E.g. `\(f : integer -> integer) -> f` is a higher-order builtin, but it's definable and is clearly constant-cost, so there's nothing fancy about costing it.

I think the person giving you comments on higher-order builtins was likely thinking about applying functions and actually reducing them from within a builtin, which we do not support indeed, but

1. we did support that in the past
2. we'd never costed that functionality, but given that it was simply continuing evaluation before returning from the builtin using the same evaluator without any state reinitialization or discarding, I don't think there were any costing issues either, it just didn't matter if the CEK machine (or any other evaluator) execution was invoked directly or continued from within a builtin

> so I was wondering if there is any problem here.

Anyhow, take my words with a grain of salt, I might be failing to spot an issue with costing pattern matching builtins.